### PR TITLE
[MoveLib] Add Delegation, reward distribution

### PR DESCRIPTION
**Pull request by lxfind**
This PR adds the following things to the governance contracts:
Delegation: a SUI token holder can delegate their stake to an individual validator. They can claim rewards from each epoch individually latter.
Reward distribution: at end of epoch, certain amount of storage charge and computation charge are passed in and distributed proportionally.

TODO: will start adding more comments and tests while this is being reviewed.

Things not covered in this PR:
1. The reward distribution is rather too simple at the moment. Will need to introduce a few parameters later.
2. Validators don't charge fees yet.

### CODE REVIEW

File: sui_programmability/framework/sources/Governance/ValidatorSet.move [hunk 1]
```diff
@@ -265,9 +329,17 @@ module Sui::ValidatorSet {
...
-    #[test_only]
-    public fun get_total_stake(self: &ValidatorSet): u64 {
-        self.total_stake
+    fun distribute_reward(validators: &mut vector<Validator>, total_stake: u64, reward: &mut Coin<SUI>, ctx: &mut TxContext) {
+        let total_reward_amount = Coin::value(reward);
+        let length = Vector::length(validators);
+        let i = 0;
+        while (i < length) {
+            let validator = Vector::borrow_mut(validators, i);
+            let reward_amount = Validator::get_stake_amount(validator) * total_reward_amount / total_stake;
```
Comment (tnowacki) [inline sui_programmability/framework/sources/Governance/ValidatorSet.move:338]: I don't know enough about the tokenomics math here, but it feels like there should be some assertion about the `Coin::value(reward)` at the end. 
Mostly this division math makes me uncomfortable (in that there could be some unexpected remainder at the end) 

If no need for an assertion, maybe just a comment explaining this?

File: sui_programmability/framework/sources/Governance/Delegation.move [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    /// Checks whether the delegation object is eligible to claim the reward
+    /// given the epoch to claim and the validator address.
+    public(friend) fun can_claim_reward(
+        self: &Delegation,
+        epoch_to_claim: u64,
+        validator: address,
+    ): bool {
+        if (validator != self.validator_address) {
+            return false
+        };
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/Delegation.move:83]: I believe these guys can be chained with `else if`.

File: sui_programmability/framework/sources/Governance/Delegation.move [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        let stake = Option::extract(&mut self.active_delegation);
+        let sender = TxContext::sender(ctx);
+        Transfer::transfer(stake, sender);
+
+        self.ending_epoch = Option::some(ending_epoch);
+    }
+
+    /// Checks whether the delegation object is eligible to claim the reward
+    /// given the epoch to claim and the validator address.
+    public(friend) fun can_claim_reward(
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/Delegation.move:76]: Does this function need to be a `friend` if it doesn't mutate state? It's not important, but still. Maybe someone could write some helper/wrapper for delegation?

File: sui_programmability/framework/sources/Governance/Delegation.move [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            next_reward_unclaimed_epoch: starting_epoch,
+            validator_address,
+        };
+        Transfer::transfer(delegation, TxContext::sender(ctx))
+    }
+
+    /// Deactivate the delegation. Send back the stake and set the ending epoch.
+    public(friend) fun undelegate(
+        self: &mut Delegation,
+        ending_epoch: u64,
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/Delegation.move:62]: Is there any validation on the `ending_epoch`?

File: sui_programmability/framework/sources/Governance/Delegation.move [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        validator_address: address,
+    }
+
+    public(friend) fun create(
+        starting_epoch: u64,
+        validator_address: address,
+        stake: Coin<SUI>,
+        ctx: &mut TxContext,
+    ) {
+        let delegate_amount = Coin::value(&stake);
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/Delegation.move:47]: Can the value be `0`?

File: sui_programmability/framework/sources/Governance/Delegation.move [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        }
+    }
+
+    /// Claim delegation reward. Increment next_reward_unclaimed_epoch.
+    public(friend) fun claim_reward(
+        self: &mut Delegation,
+        reward: Coin<SUI>,
+        ctx: &mut TxContext,
+    ) {
+        let sender = TxContext::sender(ctx);
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/Delegation.move:98]: Super nit: but I think here and everywhere passing `TxContext::sender()` directly is okay. It's like a super common thing and everyone knows that it means sender by the look of it.

File: sui_programmability/framework/sources/Governance/Delegation.move [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+            next_reward_unclaimed_epoch,
+            validator_address: _,
+        } = self;
+        ID::delete(id);
+        Option::destroy_none(active_delegation);
+        let ending_epoch = *Option::borrow(&ending_epoch);
+        assert!(next_reward_unclaimed_epoch == ending_epoch, 0);
+    }
+
+    public(script) fun tranfser(self: Delegation, recipient: address, _ctx: &mut TxContext) {
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/Delegation.move:123]:
```suggestion
    public(script) fun transfer(self: Delegation, recipient: address, _ctx: &mut TxContext) {
```

File: sui_programmability/framework/sources/Governance/EpochRewardRecord.move [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        Transfer::freeze_object(EpochRewardRecord {
+            id: TxContext::new_id(ctx),
+            epoch,
+            computation_charge,
+            total_stake,
+            validator,
+        })
+    }
+
+    public fun get_epoch(self: &EpochRewardRecord): u64 {
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/EpochRewardRecord.move:41]: I believe it is common to name getters after properties:
```suggestion
    public fun epoch(self: &EpochRewardRecord): u64 {
```

And same applies below.

File: sui_programmability/framework/sources/Governance/SuiSystem.move [hunk 9]
```diff
@@ -73,6 +78,7 @@ module Sui::SuiSystem {
                 max_validator_stake,
                 max_validator_candidate_count,
             },
+            delegation_reward: Coin::zero(ctx),
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/SuiSystem.move:81]: This is going to be sooo much cooler when we add Balances. 😆

File: sui_programmability/framework/sources/Governance/SuiSystem.move [hunk 10]
```diff
@@ -153,21 +159,94 @@ module Sui::SuiSystem {
         )
     }
 
+    public(script) fun request_add_delegation(
+        self: &mut SuiSystemState,
+        delegate_stake: Coin<SUI>,
+        validator_address: address,
+        ctx: &mut TxContext,
+    ) {
+        let amount = Coin::value(&delegate_stake);
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/SuiSystem.move:168]: How do we make sure that amounts are valid (ie non zero)? Is there like a protection or it is okay to pass `Coin::zero()` inside?

File: sui_programmability/framework/sources/Governance/ValidatorSet.move [hunk 11]
```diff
@@ -159,10 +200,21 @@ module Sui::ValidatorSet {
...
 
-        self.total_stake = calculate_total_stake(&self.active_validators);
+        let (validator_stake, delegation_stake) = calculate_total_stake(&self.active_validators);
+        self.validator_stake = validator_stake;
+        self.delegation_stake = delegation_stake;
+
         self.quorum_threshold_pct = calculate_quorum_threshold(&self.active_validators);
     }
 
+    public fun get_validator_stake(self: &ValidatorSet): u64 {
```
Comment (damirka) [inline sui_programmability/framework/sources/Governance/ValidatorSet.move:210]:
```suggestion
    public fun validator_stake(self: &ValidatorSet): u64 {
```

**Review by damirka — COMMENTED** [review]
I need another take to go through logic. 
So far only commented on code style and data validation.

File: sui_programmability/framework/sources/Coin.move [hunk 12]
```diff
@@ -116,6 +118,18 @@ module Sui::Coin {
         Coin { id: TxContext::new_id(ctx), value }
     }
 
+    /// A special function that can mint coin without increasing the supply
+    /// in the treasury cap. This can only be called by SuiSystem (friend) to mint
+    /// staking reward at epoch boundaries. This is safe because the reward
+    /// comes from gas charges during transactions.
+    public(friend) fun mint_sui_reward<T>(
```
[the remark this answers is not in the document]
Comment (lxfind) [reply sui_programmability/framework/sources/Coin.move:125]: `SuiSystem` does have `TreasuryCap<SUI>`. (in fact, to call `mint_sui_reward`, `SuiSystem` needs to pass the cap as argument for double safety).
The problem here is that we don't want to increase the supply in this special case.

**Comment by lxfind** [comment]
Added counter to track the number of delegators who haven't claimed the reward. This will allow us to delete the record object once everyone has done so.

File: sui_programmability/framework/sources/Governance/ValidatorSet.move [hunk 13]
```diff
@@ -16,8 +17,12 @@ module Sui::ValidatorSet {
     friend Sui::ValidatorSetTests;
 
     struct ValidatorSet has store {
-        /// Total amount of stake from all active validators, at the beginning of the epoch.
-        total_stake: u64,
+        /// Total amount of stake from all active validators (not include delegation),
```
Comment (sblackshear) [inline sui_programmability/framework/sources/Governance/ValidatorSet.move:20]:
```suggestion
        /// Total amount of stake from all active validators (not including delegation),
```

**Review by sblackshear — APPROVED** [review]

**Pull request merged**